exchanges/okx/okx_wrapper.go

```diff
@@ -0,0 +1,1217 @@
+		})
+	case asset.Option:
+		var instsb []Instrument
+		var instsc []Instrument
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "BTC-USD",
+		})
+		if er != nil {
+			goto checkErrorAndContinue
+		}
+		instsb, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "ETH-USD",
+		})
+		if er != nil {
+			goto checkErrorAndContinue
+		}
+		instsc, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeOption,
+			Underlying:     "SOL-USD",
+		})
+		insts = append(insts, instsb...)
+		insts = append(insts, instsc...)
+	case asset.Margin:
+		insts, er = ok.GetInstruments(ctx, &InstrumentsFetchParams{
+			InstrumentType: okxInstTypeMargin,
+		})
+	}
+checkErrorAndContinue:
+	if er != nil || len(insts) == 0 {
+		return pairs, er
+	}
+	for x := range insts {
+		var pair string
+		switch insts[x].InstrumentType {
+		case asset.Spot:
+			pair = insts[x].BaseCurrency + format.Delimiter + insts[x].QuoteCurrency
+		case asset.Futures, asset.PerpetualSwap, asset.Option:
+			currency, err := currency.NewPairFromString(insts[x].Underlying)
+			if err != nil {
+				continue
+			}
+			pair = currency.Base.String() + format.Delimiter + currency.Quote.String()
+		}
+		pairs = append(pairs, pair)
+	}
+	return pairs, nil
+}
+
+// UpdateTradablePairs updates the exchanges available pairs and stores them in the exchanges config
+func (ok *Okx) UpdateTradablePairs(ctx context.Context, forceUpdate bool) error {
+	assetTypes := ok.GetAssetTypes(false)
+	for i := range assetTypes {
+		if !(assetTypes[i] == asset.Spot || assetTypes[i] == asset.PerpetualSwap || assetTypes[i] == asset.Option || assetTypes[i] == asset.Futures || assetTypes[i] == asset.Margin) {
+			return errInvalidInstrumentType
+		}
+		p, err := ok.FetchTradablePairs(ctx, assetTypes[i])
+		if err != nil {
+			return err
+		}
+		selectedPairs := []string{}
+		pairsMap := map[string]int{}
+		for i := range p {
+			if p[i] == "" {
+				continue
+			}
+			count, ok := pairsMap[p[i]]
+			if !ok || count == 0 {
+				pairsMap[p[i]] = 1
+				selectedPairs = append(selectedPairs, p[i])
+			} else {
+				continue
+			}
+		}
+		pairs, err := currency.NewPairsFromStrings(selectedPairs)
+		if err != nil {
+			return err
+		}
+		err = ok.UpdatePairs(pairs, assetTypes[i], false, forceUpdate)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// UpdateTicker updates and returns the ticker for a currency pair
+func (ok *Okx) UpdateTicker(ctx context.Context, p currency.Pair, a asset.Item) (*ticker.Price, error) {
+	var mdata *TickerResponse
+	var er error
+	var instrumentID string
+	instrumentID, er = ok.getInstrumentIDFromPair(p, a)
+	if er != nil {
+		return nil, er
+	}
+	switch a {
+	case asset.Spot, asset.Margin, asset.PerpetualSwap, asset.Futures, asset.Option:
+		mdata, er = ok.GetTicker(context.Background(), instrumentID)
```

Comment:
please use the context from the parameter.

Reply:
What about to non-REST HTTP call methods that don't have a context in their parameter?

Reply:
It's better to populate the context into the parameter of the non-REST http calls. Then for example in websocket, we can just do a context.TODO(). In future context will be populated through the websocket system.